Allow decorators in JavaScript files

billti opened the pull request:
Fixes #6872

mhegazy commented on src/compiler/program.ts:
```diff
@@ -1168,9 +1168,6 @@ namespace ts {
                             let typeAssertionExpression = <TypeAssertion>node;
                             diagnostics.push(createDiagnosticForNode(typeAssertionExpression.type, Diagnostics.type_assertion_expressions_can_only_be_used_in_a_ts_file));
                             return true;
-                        case SyntaxKind.Decorator:
```
since this is a non-standard feature, we should ask you to set a flag.. i.e. error if the flag is not set, and ask the user to specify the flag.

billti replied:
Hmmmm. I'm conflicted on this one, as it seems maybe a little user hostile. (Kind of like we used to require a 'module' setting for you to write a module). Its not like they are going to write valid decorator syntax by accident, they opt in to decorators by virtue of writing one, and there is only one setting for the config option (`true`), but we'd go make them add a config file and the option anyway, (even though for many Salsa users they are just editing code, not compiling).

On the other hand, I get that the feature is "experimental" and we have precedent. Let's chat tomorrow, and I'll try and prepare the change anyway in the meantime. ( @egamma Any strong thoughts?).

mhegazy replied:
i am just worried of breaking these users later with no warning. the switch is a implicit agreement that you may be broken. not sure if this changes your feelings when you really are broken though.

egamma replied:
My only strong feeling is that showing an error is too much, showing a warning that this is an experimental feature would be more appropriate. But then the warning has to tell me how to get rid of it.

billti commented:
I've updated this to give the same warning in JavaScript that it gives in TypeScript, namely to set `experimentalDecorators`. Note I changed the error message slightly also, as a JavaScript user is unlikely to be using the command-line compiler, but will want to know they need to *set* the option (in the ?sconfig.json file).

I can port this to `release-1.8` also if you think this is low risk enough. (Seems desirable for Salsa).

DanielRosenwasser commented on src/compiler/diagnosticMessages.json:
```diff
@@ -687,7 +687,7 @@
         "code": 1218
     },
-    "Experimental support for decorators is a feature that is subject to change in a future release. Specify '--experimentalDecorators' to remove this warning.": {
+    "Experimental support for decorators is a feature that is subject to change in a future release. Set 'experimentalDecorators' to remove this warning.": {
```
"the 'experimentalDecorators' option"

DanielRosenwasser commented on tests/baselines/reference/generatorTypeCheck39.errors.txt:
```diff
@@ -13,7 +13,7 @@ tests/cases/conformance/es6/yieldExpressions/generatorTypeCheck39.ts(7,13): erro
 !!! error TS1163: A 'yield' expression is only allowed in a generator body.
         class C {
               ~
```
It's kind of weird that we error on the node instead of the decorator itself. We should consider that in a separate change.

DanielRosenwasser commented on tests/cases/fourslash/getJavaScriptSemanticDiagnostics21.ts:
```diff
@@ -1,15 +1,8 @@
-  }
-]`);
\ No newline at end of file
+verify.getSemanticDiagnostics(`[]`);
```
Consider just `verify.numberOfErrorsInCurrentFile(0);`

DanielRosenwasser commented:
Looks good. Just get the tests working (see #7069).

billti commented:
@mhegazy Any thoughts? Else I'll port this to 1.8 also before Tuesday.

I do think ideally the error logic should probably be in the same location for TypeScript & JavaScript, (and probably not in the type checker as its really a syntax issue), but we could look at refactoring this as part of #6802.

mhegazy commented:
sorry for the delay. :+1:

RyanCavanaugh commented:
:+1:
